flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,614 @@
+      var rng: Random = _
+      var fraction: Double = _
+
+      override def open(parameters: Configuration): Unit ={
+        rng = new Random()
+        fraction = getRuntimeContext.getBroadcastVariable(RANDOM_FRACTION).get(0)
+      }
+      override def filter(value: (Vector,Int)): Boolean = {
+        rng.nextDouble() < fraction * value._2
+      }}
+    ).withBroadcastSet(fraction, RANDOM_FRACTION).map(x => x._1)
+
+    data.getExecutionEnvironment.fromElements(k).mapWithBcSet(extraSampledSet){
+      (required, largeSample) =>
+        val output = Array.ofDim[LabeledVector](required)
+        for(i<- 0 to required - 1){
+          output(i) = LabeledVector(i, largeSample(i))
+        }
+        output.toSeq
+    }
+  }
+
+  /**
+   * Pick k centers from data one by one using kmeans++ initialization scheme
+   *
+   * The k-means++ scheme works as described by the original authors in
+   * [[http://ilpubs.stanford.edu:8090/778/1/2006-13.pdf]]
+   *
+   * Given a data set X with |X| points, the k-means++ algorithm proceeds as follows:
+   *
+   * 1. Initialize C \leftarrow \{\}
+   * 2. Let p be a point sampled uniformly at random from X. C \leftarrow C \cup \{p\}
+   * 3. for i \leftarrow 1 to k-1
+   * Choose an x in X with probability \frac{d(x,C)}{sigma_nolimits{p \in X} d(p,C)} where
+   * d(p,C) denotes the distance of p from its nearest center in C
+   * 4. Output C as the initial seed points. These can now be used in a KMeans algorithm as initial
+   * centers
+   *
+   * @param data Training data set
+   * @param k Number of clusters
+   * @param blocks Blocks of data
+   * @return Initial centroids
+   */
+  private def kmeans(
+      data: DataSet[(Vector, Int)],
+      k: Int,
+      blocks: Int)
+    : DataSet[Seq[LabeledVector]] = {
+    // first pick one center randomly
+    val initialCentroids = random(data, 1)
+    initialCentroids.iterate(k - 1){
+      currentCentroids => {
+        // sample one point from each block based on the local probability distribution
+        val blockSamples = FlinkMLTools.block(data, blocks, Some(ModuloKeyPartitioner))
+          .mapWithBcVariable(currentCentroids) {
+          (block, centroids) => {
+            val rng = new Random()
+            // form a cumulative distribution
+            val distances = Array.ofDim[Double](block.values.length)
+            distances(0) =
+              findNearestCentroid(block.values.head._1, centroids)._2 * block.values.head._2
+            for (i <- 1 to block.values.length - 1) {
+              distances(i) = distances(i - 1) +
+                findNearestCentroid(block.values(i)._1, centroids)._2 * block.values(i)._2
+            }
+            val samplePoint = sampleFromDistribution(rng.nextDouble() * distances.last, distances)
+            (block.values(samplePoint)._1, distances.last)
+          }
+        }
+        // now sample one point from the block sample
+        currentCentroids.mapWithBcSet(blockSamples) {
```

Comment:
Can't we sample the point from the `blockSamples` by calling a reduce operation on it which receives the `currentCentroids` as the broadcast set?